rust/sedona-datasource/src/format.rs

```diff
@@ -0,0 +1,717 @@
+}
+
+impl ExternalFileSource {
+    pub fn new(spec: Arc<dyn ExternalFormatSpec>) -> Self {
+        Self {
+            spec,
+            batch_size: None,
+            file_schema: None,
+            file_projection: None,
+            filters: Vec::new(),
+            metrics: ExecutionPlanMetricsSet::default(),
+            projected_statistics: None,
+        }
+    }
+}
+
+impl FileSource for ExternalFileSource {
+    fn create_file_opener(
+        &self,
+        store: Arc<dyn ObjectStore>,
+        base_config: &FileScanConfig,
+        _partition: usize,
+    ) -> Arc<dyn FileOpener> {
+        let args = OpenReaderArgs {
+            src: Object {
+                store: Some(store.clone()),
+                url: Some(base_config.object_store_url.clone()),
+                meta: None,
+                range: None,
+            },
+            batch_size: self.batch_size,
+            file_schema: self.file_schema.clone(),
+            file_projection: self.file_projection.clone(),
+            filters: self.filters.clone(),
+        };
+
+        Arc::new(ExternalFileOpener {
+            spec: self.spec.clone(),
+            args,
+        })
+    }
+
+    fn try_pushdown_filters(
+        &self,
+        filters: Vec<Arc<dyn PhysicalExpr>>,
+        _config: &ConfigOptions,
+    ) -> Result<FilterPushdownPropagation<Arc<dyn FileSource>>> {
+        // Record any new filters
+        let num_filters = filters.len();
+        let mut new_filters = self.filters.clone();
+        new_filters.extend(filters);
+        let source = Self {
+            filters: new_filters,
+            ..self.clone()
+        };
+
+        // ...but don't indicate that we handled them so that the filters are
+        // applied by the other node.
+        Ok(FilterPushdownPropagation::with_parent_pushdown_result(vec![
+            PushedDown::No;
+            num_filters
+        ])
+        .with_updated_node(Arc::new(source)))
+    }
+
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    fn with_batch_size(&self, batch_size: usize) -> Arc<dyn FileSource> {
+        Arc::new(Self {
+            batch_size: Some(batch_size),
+            ..self.clone()
+        })
+    }
+
+    fn with_schema(&self, schema: SchemaRef) -> Arc<dyn FileSource> {
+        Arc::new(Self {
+            file_schema: Some(schema),
+            ..self.clone()
+        })
+    }
+
+    fn with_projection(&self, config: &FileScanConfig) -> Arc<dyn FileSource> {
+        Arc::new(Self {
+            file_projection: config.file_column_projection_indices(),
+            ..self.clone()
+        })
+    }
+
+    fn with_statistics(&self, statistics: Statistics) -> Arc<dyn FileSource> {
+        Arc::new(Self {
+            projected_statistics: Some(statistics),
+            ..self.clone()
+        })
+    }
+
+    fn metrics(&self) -> &ExecutionPlanMetricsSet {
+        &self.metrics
+    }
+
+    fn statistics(&self) -> Result<Statistics> {
```

Comment:
Why Result ? You don't use the Err but panic with `.expect()`.
IMO it would be better to return `Option<Statistics>` instead.